hasona23 (hm) Habit Logger project

hasona23 opened the pull request:
dear sir,
this is the habit logger project

chrisjamiecarter commented on HabitTracker/HabitTracker/Program.cs:
```diff
@@ -0,0 +1,298 @@
+            name = Console.ReadLine();
+        } while (!HabitDatabase.HabitTypes.ContainsKey(name) && !string.IsNullOrEmpty(name));
+
+
```
Code Conventions: Spacing. Redundant blank line (no double blank lines).

chrisjamiecarter commented on HabitTracker/HabitTracker/Program.cs:
```diff
@@ -0,0 +1,298 @@
+            }
+        } while (true);
+
+        Console.WriteLine("Enter habit's amount ##-" + HabitDatabase.HabitTypes[name].Unit);
```
Use interpolated strings here:
❌ `Console.WriteLine("Enter habit's amount ##-" + HabitDatabase.HabitTypes[name].Unit);`
✔️ `Console.WriteLine($"Enter habit's amount ##-{HabitDatabase.HabitTypes[name].Unit}");`

chrisjamiecarter commented on HabitTracker/HabitTracker/Program.cs:
```diff
@@ -0,0 +1,298 @@
+        Console.WriteLine("Finally create the habit's measurement.");
+        Console.WriteLine("press enter to avoid or enter it in this format ##-unit");
+
+        
```
Code Conventions: Spacing. Redundant blank line (no double blank lines).

chrisjamiecarter commented on HabitTracker/HabitTracker/Program.cs:
```diff
@@ -0,0 +1,298 @@
+            else
+                amount = read;
+        
+
```
Code Conventions: Spacing. Redundant blank line (no double blank lines).

chrisjamiecarter commented on HabitTracker/HabitTrackerLibrary/Habit.cs:
```diff
@@ -0,0 +1,72 @@
+    private readonly Amount? _amount;
+    //quantity is the optional amount to measure and displayed as "39-unit"
+
+
```
Code Conventions: Spacing. Redundant blank line (no double blank lines).

chrisjamiecarter commented on HabitTracker/HabitTrackerLibrary/Habit.cs:
```diff
@@ -0,0 +1,72 @@
+        public string Unit = "";
+        public int Value;
+
+        public Amount()
```
Is an empty constructor required here?

chrisjamiecarter commented on HabitTracker/HabitTrackerLibrary/HabitDatabase.cs:
```diff
@@ -0,0 +1,168 @@
+
+    public static void CreateDatabase()
+    {
+        using (var connection = new SqliteConnection(_connectionString))
```
Remember that you can use the simpler "Using Declaration" syntax to save an indentation level against the "Using Statement" syntax:

❌ 
```
// Using Statement
using (var connection = new SqliteConnection(connectionString))
{
    // ...
}
```

✔️ 
```
// Using Declaration
using var connection = new SqliteConnection(connectionString);
// ...
```

chrisjamiecarter commented on README.md:
This file needs to be a directory lower. Otherwise your project is not containerized and is mixed with other students.

chrisjamiecarter reviewed (changes requested):
Hey @hasona23 👋,

Congratulations and great job on another project submission! 🎉 It’s great to see your progress.

I have performed a peer review and have some praise, as well as a few suggestions:
- Well written README.md file with great content! 🏅 
- Using File-Scoped Namespaces. 🎖️ 
- Implemented DRY principle. 🥇 
- Separation of concerns. ⭐ 
- Great code styling. 🌟 
- Perfect use of parameterised sql queries. 👏 

**Solution Structure**
Re-read step 3 of the [Code Reviews](https://thecsharpacademy.com/article/52/code-reviews) academy page.
When submitting your code review pull request, all your solution files should be contained inside a new folder as per below:

✔️ 
```
.
└── [ProjectName].[GitHubUserName]/
    ├── [SolutionDirectory1]/
    ├── [SolutionDirectory2]/
    └── [ProjectName].sln
```

❌ 
You are missing the `.[GitHubUserName]` suffix.
Your README.md file needs to be inside the above folder.
This will make it easier for code-reviewers to pull down and review your code in the future.

**Issues**
I have to double key some values. For example, on add new habit:

![image](https://github.com/user-attachments/assets/26714c2f-3774-4a24-8ed3-4dc50019c3f5)

And on Delete:

![image](https://github.com/user-attachments/assets/5b46c476-2a65-4fbe-bb22-51d3d5d50880)

When adding an existing habit there is not option to allow user to input Quantity:

![image](https://github.com/user-attachments/assets/9192a1de-ac0a-41a5-be3d-427f1872fd34)

The update function doesn't seem to work at all:

![image](https://github.com/user-attachments/assets/225ea8df-626c-46f8-975a-b9d35ca51f02)

![image](https://github.com/user-attachments/assets/2d7a9459-73d5-4ae1-a0d8-da6c87827f88)

![image](https://github.com/user-attachments/assets/e379aab7-f537-44c1-8250-ac82502f450b)

I suggest revisiting to try understand and fix any issues, as you are currently failing the below requirements:

- [ ] The users should be able to insert, delete, update and view their logged habit.
- [ ] You should handle all possible errors so that the application never crashes.

Feel free to reach out if you have any questions or if you'd like to collaborate further on any of these points.

You're doing a fantastic job, your code is good, there is just some logic issues that are easily fixed! Once resolved. Please commit and push your code. This will update me and I will finish my review!

Thanks,
@chrisjamiecarter

hasona23 commented:
Hello @chrisjamiecarter ,
thank your efforts reviewing the code i addressed all the issues except for the issues except for the update function the one where running doesnt have a unit to alter  and the double key 

Details:
double key issue:
the issue was caused due to a useless Console.ReadLine() in the DatabaseUpdateType function. 

Update Function malfunction:
the update function's aim is to just edit the habits quantity not the date and the error in it was that it resumed the execution when the function doesn't have any units to change which was due to a missed return; call to stop the execution of the functions which also resulted in the NULL REFRENCE error

program crashed :
It should be fixed now i stopped the execution of function and returned to the main menu when there is not quantity  so there should be no NULL REFRENCE error

chrisjamiecarter reviewed (approved):
Perfect @hasona23 🎉,

Well done on the updates.

I still think the user experience could be greatly improved but you have fulfilled all the requirements ⭐ 

I will go ahead and mark as approved, keep up the excellent work on the next projects! 😊

Best regards,
@chrisjamiecarter
